[content-visibility] Remove scroll anchor workaround

rwlbuis opened the pull request:
#### 21d97506efba191f7fabf7affa719a646026a13b
<pre>
[content-visibility] Remove scroll anchor workaround
<a href="https://bugs.webkit.org/show_bug.cgi?id=259269">https://bugs.webkit.org/show_bug.cgi?id=259269</a>
<a href="https://rdar.apple.com/112759490">rdar://112759490</a>

Reviewed by Tim Nguyen.

Now that scroll anchoring is enabled, remove the hack of resetting all c-v: auto elements to offscreen status on scrolling.

* Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
(WebCore::ContentVisibilityDocumentState::updateContentRelevancyForScrollIfNeeded):
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::focus):

Canonical link: <a href="https://commits.webkit.org/272647@main">https://commits.webkit.org/272647@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e2de02154120bbaa5e9cb19ff53a7a74e4419316

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32422 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11161 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34254 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34979 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29340 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33277 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13515 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8365 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28875 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32846 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9412 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28979 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8195 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8341 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28911 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36315 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29473 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29340 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34464 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8469 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6420 "Found 1 new test failure: fast/replaced/pdf-as-embed-with-no-mime-type-is-not-blank.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32325 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10132 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7567 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9090 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9041 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3590,8 +3590,6 @@ void Element::focus(const FocusOptions& options)
         return;
     }
 
-    document->updateContentRelevancyForScrollIfNeeded(*this);
```
Can this function definition be removed?

rwlbuis replied:
It does not seem to make a difference (anymore?) when I remove it, looking at the wpt test results. However, if we want to play it safe, I can undo this part of the PR, let me know.

nt1m replied:
I meant, can we remove the function definition for `updateContentRelevancyForScrollIfNeeded` altogether? not just the usage of it

rwlbuis replied:
Do you mean like the first try by Nikos (this had several test fails)? :
https://github.com/WebKit/WebKit/pull/22064

Some code has to instruct the scrolled/focused to content that it is no longer skipped, otherwise the skipped content check in LocalFrameView::scrollRectToVisible will cause the operation to not go through, this is what updateContentRelevancyForScrollIfNeeded does. Let me know though if that is not what you meant.

nt1m replied:
OK sounds good, can you update the FIXME on top of `void Document::updateContentRelevancyForScrollIfNeeded(const Element& scrollAnchor)` to either point to a new bug or remove it?

It might possible there's a better place to put this code now that scroll anchoring is implemented.

rwlbuis replied:
Yeah, I was planning to remove the FIXME but forgot it before creating the new PR, will fix before landing.

nt1m reviewed (approved):
r=me modulo comment about FIXME

rwlbuis committed:
[content-visibility] Remove scroll anchor workaround
https://bugs.webkit.org/show_bug.cgi?id=259269
rdar://112759490

Reviewed by Tim Nguyen.

Now that scroll anchoring is enabled, remove the hack of resetting all c-v: auto elements to offscreen status on scrolling.

* Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
(WebCore::ContentVisibilityDocumentState::updateContentRelevancyForScrollIfNeeded):
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::focus):

Canonical link: https://commits.webkit.org/272647@main
